gossip/src/duplicate_shred_handler.rs

```diff
@@ -1,95 +1,59 @@
 use {
     crate::{
-        duplicate_shred::{into_shreds, DuplicateShred, Error},
+        duplicate_shred::{self, DuplicateShred, Error},
         duplicate_shred_listener::DuplicateShredHandlerTrait,
     },
-    log::*,
+    log::error,
     solana_ledger::{blockstore::Blockstore, leader_schedule_cache::LeaderScheduleCache},
     solana_runtime::bank_forks::BankForks,
     solana_sdk::{
         clock::{Epoch, Slot},
         pubkey::Pubkey,
     },
     std::{
+        cmp::Reverse,
         collections::{HashMap, HashSet},
         sync::{Arc, RwLock},
     },
 };
 
-const CLEANUP_EVERY_N_LOOPS: usize = 10;
 // Normally num_chunks is 3, because there are two shreds (each is one packet)
 // and meta data. So we discard anything larger than 3 chunks.
 const MAX_NUM_CHUNKS: usize = 3;
-// We only allow each pubkey to send proofs for 5 slots, because normally there
-// is only 1 person sending out duplicate proofs, 1 person is leader for 4 slots,
-// so we allow 5 here to limit the chunk map size.
-const ALLOWED_SLOTS_PER_PUBKEY: usize = 5;
-// We limit the pubkey for each slot to be 300 for now, when this limit is reached,
-// we drop 50% of pubkeys with lowest stakes. It is kept at 300 because we want
-// number of pubkeys after pruning to have roughly 2/3 of stake.
-const MAX_PUBKEY_PER_SLOT: usize = 300;
+// Limit number of entries per node.
+const MAX_NUM_ENTRIES_PER_PUBKEY: usize = 128;
```

Comment:
Originally we allow each pubkey to send in 1 proof per slot, and we allow 5 slots, so it's 5x2 = 10 entries at most (because when you have 3 chunks you'll piece together the proof and clean up chunks), why do we need to allow 128 now?

Reply:
yeah, I think 5 slots per pubkey is pretty restrictive.
In theory an honest node should be allowed to submit duplicate proofs for as many slots as it has.

Reply:
But I don't expect honest code to have 128 proofs to send, this is probably a software bug causing it to send too many duplicate proofs, if a tower is full of duplicate blocks we are in big trouble, so I would say 32 is more than enough.

Reply:
Each is for a different slot; and there is no limit how many slots are duplicate.

Reply:
Yes, that is my point, honest means this guy has a lot of stake so it's probably not intentionally doing bad things, but honest people can have buggy software, so the guy with the most stake can still suddenly send duplicate proofs for non existent future proofs, I don't think we need to allow that many. if we really have 128 future slots marked duplicate, we probably would have halted.

Reply:
Duplicate slot proofs are crucial to consensus and can indicate a malicious leader node or other issues.
The primary objective here is to identify such duplicates slots, and if there were no resource limit we should have ingested all such proofs and process them.
We can't just ignore them because the node might be buggy.